frdm_k22f support

sknsean opened the pull request:
So far I have verified i'm able to startup @ 120Mhz and use the UART.

Questions:
- Most of it is copy paste from the k64 part is that okay? Do i change the copyright to the company I'm working for? Or just add another line under the previous cp holder?
- On my TODO I have support for the Kinetis K02, but NXP does not have an eval board for that chip. How do we build test it then?

Commit:
ext: mcux: Add mcux SDK 2.5.0 for k22

Adds the mcux drivers and device header files for the k22.

Origin: NXP MCUXpresso SDK 2.5.0
URL: mcux.nxp.com
Maintained-by: External

Signed-off-by: Sean Nyekjaer <sean.nyekjaer@prevas.dk>

Commit:
ext: hal: nxp: mcux: Add option to build the SMC driver

This add an option to build the System Mode Controller,
used in K22.

Signed-off-by: Sean Nyekjaer <sean.nyekjaer@prevas.dk>

Commit:
arch: kinetis: Add support for k2x

This adds support for Kinestis K22

Signed-off-by: Sean Nyekjaer <sean.nyekjaer@prevas.dk>

Commit:
boards: arm: Add support for NXP Freedom K22 board

This adds support for NXP's Freedom K22 board.

Signed-off-by: Sean Nyekjaer <sean.nyekjaer@prevas.dk>

dbkinder commented on boards/arm/frdm_k22f/doc/frdm_k22f.rst:
```diff
@@ -0,0 +1,285 @@
+.. _frdm_k64f:
```
This looks to be a copy of the frdm_k64f.rst file, renamed as frdm_k22f.rst

The document label at the top of the file needs to be unique across all the documents, and the title (and content) below should be updated to document the FRDM-K22F board (including the image reference).

sknsean replied:
Did not plan to prioritze the documentaiton for now, but now it's fixed :-D

dbkinder commented:
As for the copyright, don't delete an existing copyright line. You can add an additional one, but we need to maintain the authoring copyright history.

sknsean commented:
ready for review

Commit:
tests/drivers: adc: add config for NXP Freedom K22F board

Add the configuration block for NXP Freedom K22F board for
the adc_api test.

Signed-off-by: Sean Nyekjaer <sean.nyekjaer@prevas.dk>

MaureenHelm commented on ext/hal/nxp/mcux/devices/MK22F51212/fsl_device_registers.h:
```diff
@@ -0,0 +1,64 @@
+/*
+ * The Clear BSD License
```
This is triggering the license failure in CI. NXP will soon release a new MCUXpresso SDK that returns to a standard BSD 3-clause license, so please use that instead when it becomes available. I'll ping you when it gets published.

sknsean replied:
Thanks :+1:

MaureenHelm replied:
New SDK was released yesterday!

MaureenHelm commented on dts/arm/nxp/nxp_k2x.dtsi:
```diff
@@ -0,0 +1,343 @@
+#include <arm/armv7-m.dtsi>
```
Add a license/copyright header

MaureenHelm commented on soc/arm/nxp_kinetis/k2x/CMakeLists.txt:
```diff
@@ -0,0 +1,4 @@
+zephyr_sources(
```
Add a license/copyright header

MaureenHelm commented on soc/arm/nxp_kinetis/k2x/dts_fixup.h:
```diff
@@ -0,0 +1,92 @@
+/* SoC level DTS fixup file */
```
Add a license/copyright header

MaureenHelm commented on boards/arm/frdm_k22f/Kconfig.board:
```diff
@@ -0,0 +1,5 @@
+
```
Add a license/copyright header

MaureenHelm commented on boards/arm/frdm_k22f/Kconfig.defconfig:
```diff
@@ -0,0 +1,107 @@
+	default y
+
+config I2C_1
+	default n
```
Remove the `default n` configs from this file, they're redundant.

MaureenHelm commented on boards/arm/frdm_k22f/doc/frdm_k22f.rst:
```diff
@@ -0,0 +1,246 @@
+   https://www.nxp.com/support/developer-resources/evaluation-and-development-boards/freedom-development-boards/mcu-boards/nxp-freedom-development-platform-for-kinetis-k22-mcus:FRDM-K22F
+
+.. _FRDM-K22F User Guide:
+   http://www.nxp.com/assets/documents/data/en/user-guides/FRDMK22FUG.pdf
```
Please update the links from http to https. See #11948

MaureenHelm commented on boards/arm/frdm_k22f/frdm_k22f.yaml:
```diff
@@ -0,0 +1,17 @@
+  - spi
+  - gpio
+  - usb_device
+  - watchdog
```
Have you tested all of these functions?

sknsean replied:
Will test and report :-)

hakehuang replied:
I got the same failure.

sknsean replied:
Great (or not) :-)
Not great for frdm_k64 either...
https://paste.pound-python.org/show/d7wTVrBF7ZhhKJjfWA7i/

MaureenHelm reviewed (changes requested):
Hi @sknsean, thanks for the board port!

> Most of it is copy paste from the k64 part is that okay?

If you've tested everything (adc, spi, i2c, etc.), yes.

> Do i change the copyright to the company I'm working for? Or just add another line under the previous cp holder?

Add another line.

> On my TODO I have support for the Kinetis K02, but NXP does not have an eval board for that chip. How do we build test it then?

Do you plan to upstream your K02 board support?

sknsean commented:
Yes I'm planning on upstreaming support for the K02.
I have modified a Freedom K22 board with a K02 :-) But NXP dosn't make an evalboard for the K02

jfischer-phytec-iot commented on soc/arm/nxp_kinetis/k2x/wdog.S:
```diff
@@ -0,0 +1,78 @@
+    bics r1, r2
+    strh r1, [r0, #WDOG_SCTRL_HI_OFFSET]
+
+    cpsie i
```
Interrupts should not be enabled here, see https://github.com/zephyrproject-rtos/zephyr/issues/11815
@MaureenHelm would it make sense to have a common wdog_enable for all kinetis devices?

sknsean replied:
I'll support the common wdog for all kinetis :-) I will be the same copy-paste in my k02 port

MaureenHelm replied:
I filed an issue #12109

MaureenHelm commented:
> I have modified a Freedom K22 board with a K02 :-) But NXP dosn't make an evalboard for the K02

Try adding a new board config inside your board folder. See `boards/arm/lpcxpresso54114/` for an example. This folder contains two board configs - one for the Cortex M4, and one for the Cortex M0+. I think you could do something similar, where you have one standard config for the K22 and a second config for the K02.

hakehuang commented:
@sknsean do you add the board readme file?

sknsean commented:
https://github.com/zephyrproject-rtos/zephyr/pull/11852/commits/a057856748add47828a0e298b880591ba46974f1#diff-da6c98e38adc5f5b538f230f4e4e6d30
this?

hakehuang commented:
there are several build errors

`
0:/home/shared/disk/zephyr_project/frdmk22f/zephyr$ ./scripts/sanitycheck -p frdm_k22f -b
JOBS: 16
Cleaning output directory /home/shared/disk/zephyr_project/frdmk22f/zephyr/sanity-out
Building testcase defconfigs...
185 tests selected, 79882 tests discarded due to filters
total complete:   67/ 185  36%  failed:    0

frdm_k22f                 samples/net/zperf/test_netusb_rndis                FAILED: build_error
        see: sanity-out/frdm_k22f/samples/net/zperf/test_netusb_rndis/build.log

total complete:   70/ 185  37%  failed:    1

frdm_k22f                 samples/net/http_server/test_usbnet_ecm            FAILED: build_error
        see: sanity-out/frdm_k22f/samples/net/http_server/test_usbnet_ecm/build.log

total complete:   72/ 185  38%  failed:    2

frdm_k22f                 samples/net/zperf/test_netusb_ecm                  FAILED: build_error
        see: sanity-out/frdm_k22f/samples/net/zperf/test_netusb_ecm/build.log

total complete:   83/ 185  44%  failed:    3

frdm_k22f                 samples/subsys/shell/shell_module/test_minimal     FAILED: build_error
        see: sanity-out/frdm_k22f/samples/subsys/shell/shell_module/test_minimal/build.log

total complete:   91/ 185  49%  failed:    4

frdm_k22f                 samples/net/http_server/test_usbnet_eem            FAILED: build_error
        see: sanity-out/frdm_k22f/samples/net/http_server/test_usbnet_eem/build.log

total complete:  131/ 185  70%  failed:    5

frdm_k22f                 samples/subsys/shell/shell_module/test             FAILED: build_error
        see: sanity-out/frdm_k22f/samples/subsys/shell/shell_module/test/build.log

total complete:  156/ 185  84%  failed:    6

frdm_k22f                 tests/drivers/uart/uart_basic_api/peripheral.uart.shell FAILED: build_error
        see: sanity-out/frdm_k22f/tests/drivers/uart/uart_basic_api/peripheral.uart.shell/build.log

total complete:  165/ 185  89%  failed:    7

frdm_k22f                 samples/net/zperf/test_netusb_eem                  FAILED: build_error
        see: sanity-out/frdm_k22f/samples/net/zperf/test_netusb_eem/build.log

total complete:  185/ 185  100%  failed:    8
177 of 185 tests passed with 0 warnings in 313 seconds
`

hakehuang reviewed (changes requested):
the build error shall be fixed / filter out

sknsean commented:
Fixed...
zephyr,shell-uart = &uart1; was added after this PR was made ;-)

dbkinder reviewed (approved):
doc additions LGTM, thanks.

Commit:
ext: mcux: Add mcux SDK 2.5.0 for k02

Adds the mcux drivers and device header files for the k02.

Origin: NXP MCUXpresso SDK 2.5.0
URL: mcux.nxp.com
Maintained-by: External

Signed-off-by: Sean Nyekjaer <sean.nyekjaer@prevas.dk>

Commit:
arch: kinetis: Add support for k0x

This adds support for Kinestis K02

Signed-off-by: Sean Nyekjaer <sean.nyekjaer@prevas.dk>

MaureenHelm commented on soc/arm/nxp_kinetis/k2x/Kconfig.defconfig.mk22f12:
```diff
@@ -0,0 +1,107 @@
+endif # PWM
+
+config SPI
+	default n
```
These two lines can be removed. Bool configs are default n unless explicitly set otherwise.

MaureenHelm commented on soc/arm/nxp_kinetis/k2x/Kconfig.series:
```diff
@@ -0,0 +1,16 @@
+	select CPU_CORTEX_M4
+	select SOC_FAMILY_KINETIS
+	select SYS_POWER_LOW_POWER_STATE_SUPPORTED
+	select CPU_HAS_SYSTICK
```
Please add `select CPU_HAS_NXP_MPU` as in `soc/arm/nxp_kinetis/k6x/Kconfig.series`

sknsean replied:
K22P121M120SF7RM.pdf mentions nothing about a memory protection unit...
The Kinetis K22 MCU Family Block Diagram have it as optional for K22

MaureenHelm commented on soc/arm/nxp_kinetis/k2x/soc.c:
```diff
@@ -0,0 +1,210 @@
+ */
+static ALWAYS_INLINE void clkInit(void)
+{
+/* Set HSRUN power mode */
```
Fix the alignment

MaureenHelm commented on soc/arm/nxp_kinetis/k2x/soc.c:
```diff
@@ -0,0 +1,210 @@
+	 * protection unit (MPU), specified by the architecture (PMSAv7), in the
+	 * Cortex-M4 core.  Instead, the processor includes its own MPU module.
+	 */
+#if defined(SYSMPU)
```
Why is this extra condition here compared to `soc/arm/nxp_kinetis/k6x/soc.c`?

sknsean replied:
There is no SYSMPU register in the HAL layer for the k22, but since it's listed as an option on nxp.com so I have kept the functionality

MaureenHelm commented on boards/arm/frdm_k22f/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+if(CONFIG_PINMUX_MCUX)
```
Please add a license/copyright header

MaureenHelm commented on boards/arm/frdm_k22f/Kconfig.board:
```diff
@@ -0,0 +1,11 @@
+#
+
+config BOARD_FRDM_K22F
+	bool "Freescale FRDM-K22F"
```
s/Freescale/NXP/

MaureenHelm commented on boards/arm/frdm_k22f/Kconfig.defconfig:
```diff
@@ -0,0 +1,104 @@
+	default n
+
+config SPI_2
+	default n
```
You can remove SPI_1 and SPI2 since they're default n.

MaureenHelm commented on boards/arm/frdm_k22f/dts_fixup.h:
```diff
@@ -0,0 +1,35 @@
+ */
+
+#define DT_FXOS8700_NAME				\
+	DT_NXP_KINETIS_I2C_40066000_NXP_FXOS8700_1D_LABEL
```
Don't worry about the 80-char limit in the fixup file. Please make these defines one line each instead of multiline

MaureenHelm commented on dts/arm/nxp/nxp_k0x.dtsi:
```diff
@@ -0,0 +1,292 @@
+#include <arm/armv7-m.dtsi>
```
Add license/copyright header

MaureenHelm commented on soc/arm/nxp_kinetis/k0x/Kconfig.defconfig.mk02f10:
```diff
@@ -0,0 +1,107 @@
+endif # PWM
+
+config SPI
+	default n
```
Remove these two lines

MaureenHelm commented on soc/arm/nxp_kinetis/k0x/dts_fixup.h:
```diff
@@ -0,0 +1,92 @@
+/* SoC level DTS fixup file */
```
Add license/copyright header

MaureenHelm reviewed (changes requested):
Please update `ext/hal/nxp/mcux/README` in the commits where you import SDKs.

sknsean commented on ext/hal/nxp/mcux/drivers/kinetis/CMakeLists.txt:
```diff
@@ -23,3 +23,4 @@ zephyr_sources_ifdef(CONFIG_UART_MCUX_LPSCI   fsl_lpsci.c)
 zephyr_sources_ifdef(CONFIG_UART_MCUX_LPUART  fsl_lpuart.c)
 zephyr_sources_ifdef(CONFIG_RTC_MCUX          fsl_rtc.c)
 zephyr_sources_ifdef(CONFIG_WDT_MCUX_WDOG     fsl_wdog.c)
+zephyr_sources_ifdef(CONFIG_SMC_MCUX          fsl_smc.c)
```
Later in the commit series...
soc/arm/nxp_kinetis/k0x/Kconfig.soc
soc/arm/nxp_kinetis/k2x/Kconfig.soc

sknsean commented:
@hakehuang @MaureenHelm ping :-)

sknsean commented:
ping^2

galak commented:
Ping.

sknsean commented:
Don’t have access to hw anymore...
Someone from https://github.com/prevas-dk should pick it up :)

galak commented:
@sknsean going to close the PR, and someone can re-open if they're interested in adding support for the SoC/board.
